### ISSUE TITLE
It is not allowed to push a number to the array foo in {foo: []}

### ISSUE THREAD

**Issue by doberkofler** [microsoft/TypeScript#29398]
**TypeScript Version:**  Playground

**Search Terms:** never[] 

**Code**

```ts
const foo = { foo: [] };
foo.foo.push(1);

const bar = [];
bar.push(1);
```

**Expected behavior:**
TypeScript should infer foo to be of type `{ foo: any[] }` and therefore `foo.foo.push(1);` should be allowed like `bar.push(1);` is allowed

**Actual behavior:**
Line 2: Argument of type '1' is not assignable to parameter of type 'never'.

**Playground Link:** https://www.typescriptlang.org/play/index.html#src=const%20foo%20%3D%20%7B%20foo%3A%20%5B%5D%20%7D%3B%0D%0Afoo.foo.push(1)%3B%0D%0A%0D%0Aconst%20bar%20%3D%20%5B%5D%3B%0D%0Abar.push(1)%3B%0D%0A

**Related Issues:**

**Comment by AnyhowStep** [comment]
That's really weird. I would rather both cases be inferred as `never[]`

**Comment by doberkofler** [comment]
@AnyhowStep I agree that both cases should be the same but why `never[]` and not `any[]` ?

**Comment by fatcerberus** [comment]
For `bar`, TS "evolves" the type with each `.push()`.  So it starts out as `never[]` but then evolves to `number[]`.  Push a string afterwards and it may then evolve to `(number | string)[]`, etc.  I don't believe that works with `[]` in object literals, though--after the initial definition (which as noted, is inferred as `never[]`), the type cannot change further.

**Comment by AnyhowStep** [comment]
I tend not to use `const x = [];`.
When I do need to initialize a variable to an empty array, I tend to be explicit with the data type.

So, I'd use, `const x : /*insert-data-type-here*/[] = []`

And `never[]` is assignable to `/*insert-data-type-here*/[]` because `never` is assignable to all types.

**Comment by weswigham** [comment]
I've tagged this as bug, but it might be a current design limitation that an inner field cannot be an evolving array type.

**Comment by RyanCavanaugh** [comment]
Evolving arrays are confusing enough on bare bindings, I don't think we want to expand this into properties as well. There doesn't seem to be much demand for it.
